source/extensions/clusters/redis/redis_cluster_lb.cc

```diff
@@ -55,17 +112,61 @@ RedisClusterLoadBalancer::chooseHost(Envoy::Upstream::LoadBalancerContext* conte
     return nullptr;
   }
 
-  return slot_array_->at(hash.value() % Envoy::Extensions::Clusters::Redis::MaxSlot);
+  auto shard = slot_array_->at(hash.value() % Envoy::Extensions::Clusters::Redis::MaxSlot);
+
+  auto redis_context = dynamic_cast<RedisLoadBalancerContext*>(context);
+  if (redis_context && redis_context->isReadCommand()) {
+    switch (redis_context->readPolicy()) {
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::Master:
+      return shard->master_;
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::PreferMaster:
+      if (shard->master_->health() == Upstream::Host::Health::Healthy) {
+        return shard->master_;
+      } else {
+        return chooseRandomHost(shard->all_hosts_, random_);
+      }
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::Replica:
+      return chooseRandomHost(shard->replicas_, random_);
+    case NetworkFilters::Common::Redis::Client::ReadPolicy::PreferReplica:
+      if (!shard->replicas_.healthyHosts().empty()) {
+        return chooseRandomHost(shard->replicas_, random_);
+      } else {
+        return chooseRandomHost(shard->all_hosts_, random_);
```

Comment:
Is there a reason for the lack of symmetry with respect to PreferMaster? In other words, why does the no replicas available in this scenario not return shard->master_ instead of chooseRandomHost(shard->all_hosts_, random_)?

Reply:
by doing chooseRandomHost(shard->all_hosts_, random_) will make sure that we pick a random host if the master and replicas are all unhealthy.  If only the master is healthy, then that's the host that will be picked.